app/script/calling/VideoGridRepository.js

```diff
@@ -0,0 +1,114 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ *
+ */
+
+'use strict';
+
+window.z = window.z || {};
+window.z.calling = z.calling || {};
+
+z.calling.VideoGridRepository = (() => {
+  return class VideoGridRepository {
+    /**
+     * Construct an new VideoGridRepository.
+     * @param {MediaRepository} mediaRepository - Repository for the media streams
+     * @param {CallingRepository} callingRepository - Repository for the calls
+     */
+    constructor(mediaRepository, callingRepository) {
+      const streamHandler = mediaRepository.stream_handler;
+      const streamsInfo = streamHandler.remoteMediaStreamInfoIndex.video;
+      const localMediaStream = streamHandler.localMediaStream;
+      const selfStreamState = callingRepository.selfStreamState;
+      const calls = callingRepository.calls;
+      this.grid = ko.observableArray([0, 0, 0, 0]);
+      this.selfId = ko.observable(null);
+      this.thumbnailStream = ko.observable(null);
+      this.mirrorSelf = ko.pureComputed(() => {
+        return !selfStreamState.screenSend();
+      });
+
+      const selfStream = ko.pureComputed(() => {
+        return selfStreamState.videoSend() || selfStreamState.screenSend() ? localMediaStream() : null;
+      });
+      this.selfStreamMuted = ko.pureComputed(() => !selfStreamState.audioSend());
+
+      this.streams = ko.pureComputed(() => {
+        const remoteStreams = filterUnsentStreams(streamsInfo(), calls());
+        this.selfId(selfStream() ? selfStream().id : null);
+
+        if (remoteStreams.length === 1) {
+          this.thumbnailStream(selfStream());
+          return remoteStreams;
+        }
+        this.thumbnailStream(null);
+        return selfStream() ? remoteStreams.concat(selfStream()) : remoteStreams;
+      });
+
+      this.streams.subscribe(this.updateGrid.bind(this));
+      this.updateGrid(this.streams());
+    }
+
+    /**
+     * Will compute the next grid layout according to the previous state and the new array of streams
+     * The grid will fill according to this pattern
+     * - 1 stream : [id, 0, 0, 0]
+     * - 2 streams: [id, 0, id, 0]
+     * - 3 streams: [id, 0, id, id]
+     * - 3 streams: [id, id, 0, id]
+     * - 4 streams: [id, id, id, id]
+     * @param {Array<string|0>} previousGrid - the previous state of the grid
+     * @param {Array<MediaStream>} streams - the new array of streams to dispatch in the grid
+     *
+     * @returns {Array<string|0>} the new grid
+     */
+    computeGrid(previousGrid, streams) {
+      const previousStreamIds = previousGrid.filter(streamId => streamId !== 0);
+      const currentStreamIds = streams.map(participant => participant.id);
+
+      const addedStreamIds = z.util.ArrayUtil.getDifference(previousStreamIds, currentStreamIds);
+
+      const filteredGrid = previousGrid.map(id => (currentStreamIds.includes(id) ? id : 0));
```

Comment:
`const filteredGrid = previousGrid.map(id => (currentStreamIds.includes(id) ? id : 0));` ->`const filteredGrid = previousGrid.map(id => currentStreamIds.includes(id) ? id : 0);`

Reply:
I fully agree, but prettier likes it the other way :/